File: py/0208_ImplementTriePrefixTree.py
# 208. Implement Trie (Prefix Tree)


class TrieNode:
    def __init__(self):
        self.mapping = {}
        self.isWord = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root
        for c in word:
            if c not in curr.mapping:
                curr.mapping[c] = TrieNode()
            curr = curr.mapping[c]
        curr.isWord = True

    def search(self, word: str) -> bool:
        curr = self.root
        for c in word:
            if c not in curr.mapping:
                return False
            curr = curr.mapping[c]
        return curr.isWord

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for c in prefix:
            if c not in curr.mapping:
                return False
            curr = curr.mapping[c]
        return True


if __name__ == "__main__":
    puzzles = [
        (["Trie", "insert", "search", "search", "startsWith", "insert", "search"],
         [[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]),
    ]
    for words, prefixes in puzzles:
        obj = Trie()
        for word, prefix in zip(words[1:], prefixes[1:]):
            print(getattr(obj, word)(prefix[0]))
        print()

File: go/0208/0208_ImplementTriePrefixTree.go
// 208. Implement Trie (Prefix Tree)

package main

import (
	"fmt"
)

type Trie struct {
	children map[rune]*Trie
	isWord   bool
}

func Constructor() Trie {
	return Trie{children: make(map[rune]*Trie)}
}

func (t *Trie) Insert(word string) {
	curr := t
	for _, ch := range word {
		if _, exists := curr.children[ch]; !exists {
			curr.children[ch] = &Trie{children: make(map[rune]*Trie)}
		}
		curr = curr.children[ch]
	}
	curr.isWord = true
}

func (t *Trie) Search(word string) bool {
	curr := t
	for _, ch := range word {
		if _, exists := curr.children[ch]; !exists {
			return false
		}
		curr = curr.children[ch]
	}
	return curr.isWord
}

func (t *Trie) StartsWith(prefix string) bool {
	curr := t
	for _, ch := range prefix {
		if _, exists := curr.children[ch]; !exists {
			return false
		}
		curr = curr.children[ch]
	}
	return true
}

func main() {
	obj := Constructor()
	obj.Insert("apple")
	fmt.Println(obj.Search("apple"))   // true
	fmt.Println(obj.Search("app"))     // false
	fmt.Println(obj.StartsWith("app")) // true
	obj.Insert("app")
	fmt.Println(obj.Search("app")) // true
}
